### PR TITLE
feat(sdk): app::init codegen

### PR DESCRIPTION
**Pull request by petarjuki7**
# Initialization function in sdk

## Summary

Added #[app::init] attribute, needed for initialization of application.

Fixes https://github.com/calimero-network/core/issues/247

## Improvements

It generates code that expects you to return Self, could make it enforced during compilation (building of ast).

### CODE REVIEW

File: crates/sdk/macros/src/lib.rs
Comment (miraclx) [inline crates/sdk/macros/src/lib.rs]: A pattern you'll notice that defines the design of our macros is delegation.

This function delegates to `LogicImpl` which delegates to another parser for each individual function in the impl block.

This should've been in `LogicImpl` instead (after parsing all the `fn`s)

File: crates/sdk/macros/src/lib.rs
Comment (miraclx) [inline crates/sdk/macros/src/lib.rs]: also, it should be possible to break up `#[app::logic]` into multiple blocks, so we can't make the assumption that the absence of the `init` method in this one, means there's not one in another

would be nice if we're able to add a compile-time assertion from `#[app::state]` that `<Ty>::init` is defined

As for now, I'm not sure exactly what that would look like since `Fn<A: Tuple>` and `Tuple` is nightly-only, so we might not be able to do it without making a custom trait for all `Fn` with assumptions on parameter count.

File: crates/sdk/macros/src/logic.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic.rs]: this is not a fatal parse error, collect it and proceed (one of the design considerations is to be as exhaustive as possible in a single pass)

File: crates/sdk/macros/src/logic.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic.rs]: why not leave this to `LogicMethod`?

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]: why not merge this with `PublicLogicMethod`?

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]: ah, neat trick.. helps avoid the struct definition

File: crates/sdk/macros/src/lib.rs
Comment (miraclx) [reply crates/sdk/macros/src/lib.rs]: On second thought, you can define a marker trait like `AppStateInit` that only gets implemented when a method is annotated with `#[app::init]` to guard against duplication.

And make `AppState: AppStateInit` so it requires that to be implemented.

Then we can use the `on_unimplemented` debug assertion introduced in Rust 1.78 to customize the error message.

https://blog.rust-lang.org/2024/05/02/Rust-1.78.0.html#diagnostic-attributes

**Review by miraclx — COMMENTED** [review]
PS: we also want to do away with the `.unwrap_or_default()` accommodation for `&mut self` methods.

As well as update the apps in this repo.

**Comment by petarjuki7** [comment]
> PS: we also want to do away with the `.unwrap_or_default()` accommodation for `&mut self` methods.
> 
> As well as update the apps in this repo.

Haven't added init methods yet, but changed `.unwrap_or_default()` in  `&mut self` methods.

File: crates/sdk/macros/src/lib.rs
Comment (petarjuki7) [reply crates/sdk/macros/src/lib.rs]: I'm having a problem with circular dependencies if I do `AppState: AppStateInit`, and then try to implement `AppStateInit` when I find the `#[app::init]` func  in the `macros` sdk.

I may not be doing what was intended, so feedback appreciated.

Comment (miraclx) [reply crates/sdk/macros/src/lib.rs]: You want the reverse, not this one, so - `AppState: AppStateInit` instead

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]: seems to me, this match statement is now irrelevant?

File: crates/sdk/src/state.rs
Comment (miraclx) [inline crates/sdk/src/state.rs]: can you include screenshots of this? so we know it behaves correctly?

File: crates/sdk/macros/src/logic/method.rs
Comment (petarjuki7) [reply crates/sdk/macros/src/logic/method.rs]: Should we just `.unwrap()` then? If we can guarantee that the apps state was initialized.

File: crates/sdk/macros/src/logic.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic.rs]:
```suggestion
```

File: crates/sdk/src/state.rs
Comment (petarjuki7) [reply crates/sdk/src/state.rs]: Screenshot 2024-07-25 at 17 30 27

I have the init method commented out, and we can see that the error message shows

Comment (miraclx) [reply crates/sdk/src/state.rs]: Amazing! But, please make this match the format for other types of error reports.

File: apps/gen-ext/src/lib.rs
Comment (miraclx) [inline apps/gen-ext/src/lib.rs:17]:
```suggestion
    }

```

and the others

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]:
```suggestion
pub enum Modifier<'a> {
    Init(&'a syn::Path),
```

any reason this isn't this?

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]: I think this is the same?

```suggestion
        let init_method = modifiers
            .iter()
            .any(|modifier| matches!(modifier, Modifer::Init(_)));
```

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]:
```rust
        if let (Some(_), false) = (&self.ret, init_method) {
```

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]:
```suggestion
            if init_method || matches!(&self.self_type, Some(arg::SelfType::Mutable)) {
```

or

```rust
        let state_finalizer = match (&self.self_type, init_method) {
            (Some(arg::SelfType::Mutable), _) | (_, true) => quote! {
                ::calimero_sdk::env::state_write(&app);
            },
            _ => quote! {},
        };
```

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]:
```suggestion
        let init_impl = if init_method {
```

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs:160]: let's space this out

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]: Use `errors::ParseError` and \`escape\` code references

File: crates/sdk/src/state.rs
Comment (miraclx) [inline crates/sdk/src/state.rs]:
```suggestion
pub trait Identity<This = Self> {}

impl<T: AppState> Identity<T> for T {}
```

which now satisfies the reverse constraint - `AppStateInit: AppState`

Comment (miraclx) [reply crates/sdk/src/state.rs]: Can you screenshot what it's like to implement `AppStateInit` manually?

In case they try to bypass using `#[app::init]`

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [reply crates/sdk/macros/src/logic/method.rs]: I suspect it's because it's unused, in which case you can remove the `()`

Comment (petarjuki7) [reply crates/sdk/macros/src/logic/method.rs]: Yes, it's unused, I'm going to remove the brackets

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]: another thing to check for is that the `init` method is `pub`lic

File: crates/sdk/src/state.rs
Comment (petarjuki7) [reply crates/sdk/src/state.rs]: It would throw no errors

For example if you just write:
```
impl AppStateInit for KvStore {
    type Return = KvStore;
}
```

It throws no errors

File: crates/sdk/macros/src/logic/method.rs
Comment (petarjuki7) [reply crates/sdk/macros/src/logic/method.rs]: https://github.com/calimero-network/core/blob/d107917304df78a1ee75c5d8ee7fd4702dc98cbd/crates/sdk/macros/src/logic/method.rs#L144

Is it fine because here I already make sure it is public here?

Comment (miraclx) [reply crates/sdk/macros/src/logic/method.rs]: Aha, right.. we're not bothering to parse private methods

Although, there's a concern here..

Annotating private methods with `#[app::init]` won't indicate any errors.

And you can annotate any number of them. I'll leave it to you to determine how you want to address this.

File: crates/sdk/macros/src/errors.rs
Comment (miraclx) [inline crates/sdk/macros/src/errors.rs]: perhaps we don't need to mention the macro? and since the error message is specific to init, maybe expand on the variant name? for example:

```suggestion
    #[error("an initializer, by definition, has no `self` to reference")]
    NoSelfReceiverAtInit,
```

but the final form is up to you

File: crates/sdk/src/state.rs
Comment (miraclx) [inline crates/sdk/src/state.rs]: `no method annotated`

but.. good highlight, we also need to ensure the annotated method is actually called "init"

I can't think of any good arguments for making this configurable atm

You'll also need to forbid the use of function identifiers named `init` when it's not annotated with the macro

You might ask.. why do we need the macro then? why not just check the function name to determine it's behaviour.

The idea is we don't want to leave anything implicit so while we "could" assume that your `init` method initializes the app state, you may not have had that in mind when writing the app.

So, we error on `pub fn init` with no annotation, suggesting they specify the annotation to fix it.

File: crates/sdk/src/state.rs
Comment (miraclx) [inline crates/sdk/src/state.rs:15]: this piece might be worth revisiting depending on how `rustc` displays this, since it might just be rephrasing the `message`

File: crates/sdk/src/state.rs
Comment (miraclx) [inline crates/sdk/src/state.rs]: we no longer need `Default` here

```suggestion
pub trait AppState: BorshSerialize + BorshDeserialize + AppStateInit {
```

**Comment by miraclx** [comment]
Alright, so I did some thinking.. and being able to manually implement `AppStateInit` without any init method actually being defined is unideal.

This is a non-blocker for this PR, though it's worth thinking through what a better implementation would be.

If we move the initializer method to `AppStateInit`:

```rust
pub trait AppStateInit: Sized {
    type InitParams<'a>: Deserialize<'a>;

    fn init<'a>(params: Self::InitParams<'a>) -> Self;
}
```

and since this returns an explicit `Self`, we can relinquish `Identity` of it's duty _<sub>thanks for your service</sub>_

Now, instead of code-generating an `init` method from the method, we'll move that to `#[app::state]` instead.


```rust
#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub extern "C" fn init() {
    ::calimero_sdk::env::setup_panic_hook();

    ::calimero_sdk::event::register::<#ident>();

    let Some(input) = ::calimero_sdk::env::input() else {
        ::calimero_sdk::env::panic_str("Expected input since method has arguments.")
    };

    let params = match ::calimero_sdk::serde_json::from_slice(&input) {
        Ok(value) => value,
        Err(err) => ::calimero_sdk::env::panic_str(
            &format!("Failed to deserialize input from JSON: {:?}", err)
        ),
    };

    let state = <#ident as ::calimero_sdk::state::AppStateInit>::init(params);

    ::calimero_sdk::env::state_write(&state);
}
```

With that handled, `#[app::init]` becomes the `AppStateInit` definition.

```rust
#[app::logic]
impl Adder {
    #[app::init]
    pub fn init(a: u8, b: u8) -> Adder {
        Adder { a, b }
    }

    pub fn add(&self) -> u8 {
        self.0 + self.1
    }
}
```

expanding to

```rust
impl Adder {
    pub fn init(a: u8, b: u8) -> Adder {
        Adder { a, b }
    }

    pub fn add(&self) -> u8 {
        self.0 + self.1
    }
}

const _: () = {
    struct CalimeroInput {
        a: u8,
        b: u8,
    }

    impl ::calimero_sdk::state::AppStateInit for Adder {
        type InitParams<'a> = CalimeroInput;

        fn init<'a>(
            CalimeroInput {
                a,
                b,
            }: Self::InitParams<'a>,
        ) -> Self {
            Adder::init(a, b)
        }
    }
};

#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub extern "C" fn add() {
    // Snip..
}
```

This forces implementors of `AppStateInit` to also provide the actual initialization logic.

The unfortunate downside of this is, we'll lose the ability to define generics for `#[app::state]`

Although, I suppose there's no good arguments for having them in the first place? Just that we "could".

Even though `#[app::logic]` already forbids the use of generics. So it might make sense to outlaw them for `#[app::state]` to open up this possibility.

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [reply crates/sdk/macros/src/logic/method.rs]: For posterity, the reason we use struct definitions is so you can provide a JSON payload with the field names in an object instead of just an array of parameters.

Worth noting: `serde` also attempts to deserialize an array of parameters, so we get the best of both worlds.

File: crates/sdk/src/state.rs
Comment (miraclx) [reply crates/sdk/src/state.rs]: By this, I mean prefix the tag, so users can tell which errors are `rustc` errors and which are `calimero`-flavored.

https://github.com/calimero-network/core/blob/d107917304df78a1ee75c5d8ee7fd4702dc98cbd/crates/sdk/macros/src/errors.rs#L36

**Comment by petarjuki7** [comment]
> ```rust
> pub trait AppStateInit: Sized {
>     type InitParams<'a>: Deserialize<'a>;
> 
>     fn init<'a>(params: Self::InitParams<'a>) -> Self;
> }
> ```

I was thinking of something similar, but maybe the problem is that now the user couldn't use the init method without any params.

For example how we are initializing our apps right now:
```rust
#[app::init]
pub fn init() -> KvStore {
    KvStore::default()
}
 ```
 
I agree that generics for `#[app::state]` are not needed.

**Comment by miraclx** [comment]
> but maybe the problem is that now the user couldn't use the init method without any params.

it's still possible, the macro still passes the parameters to the `AppStateInit` definition

```rust
        fn init<'a>(
            CalimeroInput {
                a,
                b,
            }: Self::InitParams<'a>,
        ) -> Self {
            Adder::init(a, b)
        }
```

File: crates/sdk/macros/src/logic/method.rs
Comment (miraclx) [inline crates/sdk/macros/src/logic/method.rs]: `Ident: PartialEq<T>` where `T: AsRef<str>` [(src)](https://docs.rs/syn/latest/syn/struct.Ident.html#impl-PartialEq%3CT%3E-for-Ident)

```rust
match (is_init, &self_type) {
    (true, Some(_)) => errors.subsume(syn::Error::new_spanned(
        input.item,
        errors::ParseError::NoSelfReceiverAtInit,
    )),
    (true, None) if name != "init" => { /* forbid this */ },
    (false, _) if name == "init" => errors.subsume(syn::Error::new_spanned(
        input.item,
        errors::ParseError::InitMethodWithoutNoInitAttribute,
    )),
    _ => {}
```

Comment (petarjuki7) [reply crates/sdk/macros/src/logic/method.rs]: Ah, okay, we are also making sure the actual function is named init
Fixed!

**Review by miraclx — APPROVED** [review]

**Comment by miraclx** [comment]
Reposted https://github.com/calimero-network/core/issues/532#issuecomment-2251864232 in #247. We'll continue the conversation there, and follow it up with it's implementation.

**Pull request merged**
